Added functions to extract dictionary for AWS cost estimate

HadleyKing opened the pull request:
RE https://github.com/biocompute-objects/galaxy/issues/28 and https://github.com/galaxyproject/galaxy/issues/9046 and https://github.com/galaxyproject/galaxy/pull/9621

    GET /api/invocations/{invocations_id}/aws_estimate
    Return a dictionary with an AWS estimate metrics for a workflow invocation.
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py
	modified:   lib/galaxy/webapps/galaxy/buildapp.py

HadleyKing committed:
Added functions to extract dictionary for AWS cost estimate

RE #28 and https://github.com/galaxyproject/galaxy/issues/9046 and https://github.com/galaxyproject/galaxy/pull/9621

    GET /api/invocations/{invocations_id}/aws_estimate
    Return a dictionary with an AWS estimate for a workflow invocation.
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py
	modified:   lib/galaxy/webapps/galaxy/buildapp.py

bgruening commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                            metrics[i].append(ds_metrics[job_output.dataset.id])
+
+        aws_dict = metrics
+        return aws_dict
```
you can return here metrics and remove the aws_dict entirely

bgruening commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                    for job_input in job.input_datasets:
+                        if hasattr(job_input.dataset, 'dataset_id'):
+                            ds_metrics[job_input.dataset.id].update({'filename': job_input.dataset.name})
+                            metrics[i].append(ds_metrics[job_input.dataset.id])
```
Can you not simply append `{'filename': job_input.dataset.name}` and remove ds_metrics?

bgruening reviewed (commented):
Maybe I'm not seeing it, but the mapping to AWS does not happen here, isn't it? This API is returning matrices about an entire workflow. So maybe the endpoints should be renamed?

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        history = workflow_invocation.history
+
+        aws_dict, metrics, ds_metrics = {}, {}, {}
+        h_contents = self.history_contents_manager.contained(history)
```
I would suggest not iterating over the history contents here, this may contain many more datasets that are not part of the workflow invocation. You can get size and other hda attributes from the job inputs and outputs

mvdbeek replied:
If you build this dictionary in the for loop below you'd also address @bgruening's comment

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        for h in h_contents:
+            ds_metrics[h.dataset.id] = {
+                'file_size': int(h.dataset.file_size),
+                'total_size': int(h.dataset.total_size),
```
For these purposes I think only `total_size` matters.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                    metrics[i] = summarize_job_metrics(trans, job)
+                    for job_input in job.input_datasets:
+                        if hasattr(job_input.dataset, 'dataset_id'):
+                            ds_metrics[job_input.dataset.id].update({'filename': job_input.dataset.name})
```
If using collections in workflows the dataset name is usually not informative. I'd suggest instead populating this with `'job_input_name': job_input.name`, which is the output parameter of the tool that corresponds to this dataset

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                            metrics[i].append(ds_metrics[job_input.dataset.id])
+                    for job_output in job.output_datasets:
+                        if hasattr(job_output.dataset, 'dataset_id'):
+                            ds_metrics[job_output.dataset.id].update({'filename': job_output.dataset.name})
```
```suggestion
                            ds_metrics[job_output.dataset.id].update({'job_output_name': job_output.name})
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        for i, step in enumerate(workflow_invocation.steps):
+            if step.workflow_step.type == 'tool':
+                for job in step.jobs:
+                    metrics[i] = summarize_job_metrics(trans, job)
```
This way you'll only collect the last jobs' metrics or a workflow step since you overwrite `metrics[i]` for each job.

I'd suggest making metrics a list of step metrics, where each metric is a dictionary describing the step and the metric. The structure could look something like
```python
[{'step_id': step.id,
  'step_label': step.label,
  jobs: [{'job_id': job.id, 'job_metrics': metrics, 'job_inputs': [], 'job_outputs': []]
}]
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,46 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        return aws_dict
+
+    @expose_api
+    def export_aws_estimate(self, trans, invocation_id, **kwd):
```
Can you rename this to `invocation_metrics` or similar ? The output may not actually contain aws estimates if that is not configured.

mvdbeek reviewed (commented):
That's a nice idea, hope my suggestions are helpful.

HadleyKing committed:
Updates to address review comments by @mvdbeek and @bgruening
Changes to be committed:
	modified:   lib/galaxy/webapps/galaxy/api/workflows.py
	modified:   lib/galaxy/webapps/galaxy/buildapp.py

HadleyKing committed:
Fix for py35-lint
Changes to be committed:
	./lib/galaxy/webapps/galaxy/api/workflows.py:1063:69: W291 trailing whitespace
	ERROR:   py35-lint

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                                    'uuid': str(h.dataset.uuid),
+                                    'job_input_name': job_input.name
+                                }
+                                metrics[i]['job_inputs'].append(h_metrics)
```
You don't need the history contents at all here:
```suggestion
                        if job_input.dataset.dataset:
                            dataset = job_input.dataset
                            dataset_metrics = {
                                'total_size': int(dataset.total_size),
                                'id': trans.security.encode_id(dataset.id),
                                'model_class': dataset.__class__.__name__,
                                'job_input_name': job_input.name
                            }
                            metrics[i]['job_inputs'].append(dataset_metrics)
```
I'd use the item id instead of the uuid. The id is the primary identifier we use in Galaxy, with this id you can get more info about this dataset for instance ... we don't support querying by uuid for all api endpoints. The input might also be a LIbraryDatasetAssociation, so that's why I'd also dump the model_class into the metrics.

You can use the same technique for job outputs.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                    job_metrics = summarize_job_metrics(trans, job)
+                    metrics[i] = {
+                        'tool_name': tool_id[0],
+                        'tool_version': tool_id[1],
```
```suggestion
                        'tool_version': job.tool_version,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                for job in step.jobs:
+                    job_metrics = summarize_job_metrics(trans, job)
+                    metrics[i] = {
+                        'tool_name': tool_id[0],
```
```suggestion
                        'tool_id': job.tool_id,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+
+        for i, step in enumerate(workflow_invocation.steps):
+            if step.workflow_step.type == 'tool':
+                tool_id = step.workflow_step.tool_id.split('/')[-2:]
```
```suggestion
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        metrics, h_metrics = {}, {}
+        h_contents = self.history_contents_manager.contained(history)
+
+        for i, step in enumerate(workflow_invocation.steps):
```
Would it make sense to make sure all steps are scheduled and have finished executing ?

HadleyKing replied:
I guess it would. I have always thought about these things in a 'historical' sense so that did not occur to me. How would you do that? Set an error if the invocation has not finished? Or set it to wait? Not sure how to do that best.

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        history = workflow_invocation.history
+
+        metrics, h_metrics = {}, {}
+        h_contents = self.history_contents_manager.contained(history)
```
```suggestion
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+                                    'total_size': int(h.dataset.total_size),
+                                    'uuid': str(h.dataset.uuid),
+                                    'job_output_name': job_output.name
+                                }
```
```suggestion
                        if job_output.dataset.dataset:
                            dataset = job_output.dataset
                            h_metrics = {
                                'total_size': int(dataset.total_size),
                                'id': trans.security.encode_id(dataset.id),
                                'model_class': dataset.__class__.__name__,
                                'job_output_name': job_output.name
                            }
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
+        """
+        decoded_workflow_invocation_id = self.decode_id(invocation_id)
+        workflow_invocation = self.workflow_manager.get_invocation(trans, decoded_workflow_invocation_id)
+        history = workflow_invocation.history
```
```suggestion
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -57,6 +58,7 @@ class WorkflowsAPIController(BaseAPIController, UsesStoredWorkflowMixin, UsesAnn
     def __init__(self, app):
         super(WorkflowsAPIController, self).__init__(app)
         self.history_manager = histories.HistoryManager(app)
+        self.history_contents_manager = history_contents.HistoryContentsManager(app)
```
```suggestion
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -21,9 +21,10 @@
 )
 from galaxy.managers import (
     histories,
+    history_contents,
```
```suggestion
```

mvdbeek reviewed (commented):
I think this is almost OK now, but are you sure you wouldn't want to walk along the invocation on the client ? I am not opposed to merging this (maybe labeling it beta), but I think there are some details here that are not captured in the metrics. subworkflow invocations are not taken into account currently, counting of initial workflow step input datasets (they are not technically produced by the invocation), discovered output collections (they don't have a JobToOutputDatasetAssociation, so this endpoint wouldn't count them).

We could implement this on the backend like you're doing here, but I think we have almost everything in place to let the client collect these things.

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Update lib/galaxy/webapps/galaxy/api/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

HadleyKing committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into galaxyproject-dev

HadleyKing committed:
Merge branch 'galaxyproject-dev' into aws_estimate
